### ISSUE TITLE
Add CI testing and official support for Python 3.10

### ISSUE THREAD

**Issue by CAM-Gerlach** [spyder-ide/qtpy#269]
Originally, part of #253 included adding CI testing, and thus official support for Python 3.10, but it turned out in PR #262 that given the conda packages for Pytest aren't officially yet declared compatible (despite the upstream PyPI packages having the 3.10 tag), it would more complex/inefficient than anticipated to add it right away, and so it was deferred to a followup.

As of now, we can either wait (hopefully <= a couple weeks) for the conda Pytest, etc packages to officially add Py3.10 support (simplest and least CI impact, but not immediate), or install the test deps (pytest and pytest-cov, etc) using separate conda/pip commands, using Python 3.10 on conda=no builds (increased CI complexity and runtime, especially with the matrix simplification and the coverage improvements, though allows us to add support right away).

I presume we want to target this for the QtPy 2.0.0 release, so while the former would be preferable, I imagine we'd want to go with the latter if it is not possible in time. @dalthviz , would it be helpful for us to open an issue on the pytest conda feedstock repo about this, and/or somewhere else? I didn't see one last time I checked.

**Comment by CAM-Gerlach** [comment]
Seems like the "waiting" approach is blocked on the cpython3.10 migration for CF, which per the discussion in conda-forge/conda-forge-pinning-feedstock#2008 may be 1-2 months at least.

So to ensure this doesn't block QtPy v2.0.0, I guess I'll just go ahead with the approach of bifurcating the `conda=yes` and `conda=no` branches in `test.sh` a bit more so the latter uses a pure `pip` test dep stack and thus can run under Python 3.10. with a bit of bash-fu in the `test.sh` I should be able to avoid appreciable CI run time impacts on either the conda=yes or no installs, and using `exclude` _should_ allow doing so with minimal extra verbosity in the matrix (unless unfixable Py3.10 problems arise), so there's little real downside to just going ahead with this, aside from a modest amount of extra time up front.

**Comment by CAM-Gerlach** [comment]
@dalthviz This is pretty much blocked on PR #268 being merged, as significant merge conflicts will result, though that PR is pretty straightforward and umabigous so I don't expect too much delay on that. It should be completely orthogonal to #270 getting addressed, minus any unforeseen issues.

**Comment by dalthviz** [comment]
Thinking about this maybe we should just wait for a bit

**Comment by CAM-Gerlach** [comment]
Sure, up to you. I've thought of ways to work around most of the downsides I initially thought in terms of complexity or CI resources with relatively little effort, aside from any issues with 3.10 itself (which we'll have to fix anyway), and it would be good to be able to declare 3.10 support in QtPy 2.0 (since it will unblock it for a lot of downstream dependents).

What were you thinking in terms of a timeline for

**Comment by dalthviz** [comment]
>What were you thinking in terms of a timeline for QtPy 2.0's release?

Well firts we will need to release v1.11.3 with the deprecation warnings. For the v2.0.0 release, we need to check the support for unscoped enum access, also would be nice to do some testing with for example Spyder as already some people did, using master to check for errors or missing things to support compatibility and maybe this issue if while working some progress is done in terms of Python 3.10 (which probably will get a bugfix release in december). Taking that into account probably a couple of weeks, maybe a month, hopefully before this year ends (that could be nice I think).

At the end, I guess from my side, the most important thing is to do a release after we did all we could think about to prevent any major issue to dependent projects on QtPy.

**Comment by CAM-Gerlach** [comment]
What would be particularly nice is having at least one CI job on linux/conda/latest Python that checks out the last stable released Spyder version and runs the Spyder test suite (or at least the non-slow tests, to avoid extra runtime, complexity and flakiness), as a proper integration/functional test on top of the existing unit tests. It should actually be pretty simple to set up; we could just use the existing Linux Spyder workflow with some minor additions and simplifications. We could also have more comprehensive workflows (that tested more platforms, versions, etc) that would only get run on demand/before release, to avoid clogging up CI bandwidth every PR.

I can open an issue if you want to discuss it further, and once #272 is merged I can go ahead with it.

**Comment by dalthviz** [comment]
>What would be particularly nice is having at least one CI job on linux/conda/latest Python that checks out the last stable released Spyder version and runs the Spyder test suite

I would say that sort of testing should be done over the Spyder side (maybe, just to have an extra check, we could setup QtPy from master as is done for other managed dependencies as QtAwesome in the Spyder testing workflows).

**Comment by CAM-Gerlach** [comment]
That certainly can be done over at the Spyder side, though it serves a somewhat different purpose, of primarily testing Spyder's compat with the latest QtPy as opposed to serving as a direct functional test of changes in QtPy itself. The idea here was to check out a known good current stable release of Spyder and ensure that QtPy changes didn't break things (and if so, it would show up right in the PR associated with the specific commit before it was merged, rather than us having to go back and manually determine what actually caused it), and that we didn't break Spyder tests in the first place. But if you don't think its a good idea, that's fine.

**Comment by dalthviz** [comment]
I think that kind of integration testing goes out of the scope of QtPy itself but thinking about it it could be nice to have a place/service to automatically do checks like that before releases (if I use certain package from his latest source how it can affect already available packages). So, that kind of automatic integration testing is a good idea but I think this is not the place to do that

**Comment by dalthviz** [comment]
Checking a little bit seems like there are already conda packages (on the conda-forge channel at least) for pytest with Python 3.10, should we maybe try again to update the CI to Python 3.10? What do you think @CAM-Gerlach @ccordoba12 ?

**Comment by ccordoba12** [comment]
I'm ok with that.

**Comment by CAM-Gerlach** [comment]
Sure, I can go ahead with it, but #295 needs to be closed or merged first (I'm reviewing it now, but TL;DR I'm a strong :-1: on it) since it will basically touch all the same lines.

**Comment by dalthviz** [comment]
> Sure, I can go ahead with it
Thanks @CAM-Gerlach :+1:
